test/functional/sqlsrv/sqlsrv_ae_output_param_sqltype_datetime.phpt

```diff
@@ -1,111 +1,161 @@
+    
+    $sqlTypeConstant = get_sqlType_constant($sqlType);
+        
+    // Call store procedure
+    $outSql = AE\getCallProcSqlPlaceholders($spname, 2);
+    
+    // Set these to NULL such that the PHP type of each output parameter is inferred
+    // from the SQLSRV_SQLTYPE_* constant
+    $c_detOut = null;
+    $c_randOut = null;
+    $stmt = sqlsrv_prepare(
+        $conn,
+        $outSql,
+        array(array( &$c_detOut, $direction, null, $sqlTypeConstant),
+        array(&$c_randOut, $direction, null, $sqlTypeConstant ))
+    );
+    if (!$stmt) {
+        die(print_r(sqlsrv_errors(), true));
+    }
+    sqlsrv_execute($stmt);
+    
+    $success = false;
+    $errors = sqlsrv_errors();
+    if (AE\IsDataEncrypted()) {
+        // With data encrypted, errors are totally expected
+        if (empty($errors)) {
+            echo "Encrypted data: $dataType should NOT be compatible with $sqlType\n";
+        } else {
+            // This should return 22018, the SQLSTATE for any incompatible conversion,
+            // except the XML type
+            $success = ($errors[0]['SQLSTATE'] === '22018');
+            if (!$success) {
+                if ($sqlType === 'SQLSRV_SQLTYPE_XML') {
+                    $success = ($errors[0]['SQLSTATE'] === '42000');
+                } else {
+                    echo "Encrypted data: unexpected errors with SQL type: $sqlType\n";
+                }
+            }
+        }
+    } else {
+        $compatible = isCompatible($dataType, $sqlType);
+        if ($compatible) {
+            if (!empty($errors)) {
+                echo "$dataType should be compatible with $sqlType.\n";
+            } else {
+                $success = true;
+            }
+        } else {
+            $implicitConv = 'Implicit conversion from data type ';
+
+            // 22018 is the SQLSTATE for any incompatible conversion errors
+            if ($errors[0]['SQLSTATE'] === '22018') {
+                $success = true;
+            } elseif (strpos($errors[0]['message'], $implicitConv) !== false) {
+                $success = true;
+            } else {
+                echo "Failed with SQL type: $sqlType\n";
+            }
+        }
+    }
+    return $success;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////
+
+$conn = AE\connect();
+
+foreach ($dataTypes as $dataType) {
+    echo "\nTesting $dataType:\n";
+    $success = true;
+
+    // create table
+    $tbname = GetTempTableName("", false);
+    $colMetaArr = array(new AE\ColumnMeta($dataType, "c_det"), new AE\ColumnMeta($dataType, "c_rand", null, false));
+    AE\createTable($conn, $tbname, $colMetaArr);
+
+    // Create a Store Procedure
+    $spname = 'selectAllColumns';
+    createProc($conn, $spname, "@c_det $dataType OUTPUT, @c_rand $dataType OUTPUT", "SELECT @c_det = c_det, @c_rand = c_rand FROM $tbname");
+
+    // insert a row
+    $inputValues = array_slice(${explode("(", $dataType)[0] . "_params"}, 1, 2);
```

Comment:
Could we have an example here of what the input looks like? Expressions like this are tricky to parse by looking.